docs on uploading attachments with traces (#519)
langchain-ai/langsmith-docs: docs/observability/how_to_guides/tracing/upload_files_with_traces.mdx

---
sidebar_position: 2
---

import {
  CodeTabs,
  PythonBlock,
  TypeScriptBlock,
} from "@site/src/components/InstructionsWithCode";

# Upload files with traces

:::tip Recommended Reading
Before diving into this content, it would be helpful to read the following guides:

- [Trace with LangSmith using the traceable decorator or wrapper](./annotate_code/#use-traceable--traceable)
  :::

> LangSmith supports uploading binary files (such as images, audio, videos, and PDFs) with your traces. This is particularly useful when working with LLM models pipelines using multimodal inputs or outputs.

Reviewer: nit: LLM models is redundant

```suggestion
LangSmith supports uploading binary files (such as images, audio, videos, and PDFs) with your traces. This is particularly useful when working with LLM pipelines using multimodal inputs or outputs.
```

              const mp3Buffer = await fs.readFile("my_mp3.mp3");
              const mp3ArrayBuffer = mp3Buffer.buffer; // Convert to ArrayBuffer

              const video = await fs.readFile("my_video.mp4"); // Uint8Array
              const pdfBuffer = await fs.readFile("my_document.pdf");
              const pdfArrayBuffer = pdfBuffer.buffer; // Convert to ArrayBuffer

              // Define example parameters
              const val = 42;
              const text = "Hello, world!";

              // Call traceableWithAttachments with the files
              const result = await traceableWithAttachments(val, text, image, mp3ArrayBuffer, video, pdfArrayBuffer);

  `,
          `In the TypeScript SDK, you can add attachments to traces by using \`Uint8Array\` or \`ArrayBuffer\` as data types.
  Each attachment's MIME type is specified within \`extractAttachments\`:\n

- \`Uint8Array\`: Useful for handling binary data directly.
- \`ArrayBuffer\`: Represents fixed-length binary data, which can be converted to \`Uint8Array\` as needed.\n
  Wrap your function with \`traceable\` and include your attachments within the \`extractAttachments\` option.

> In the TypeScript SDK, the \`extractAttachments\` function is an optional parameter in the \`traceable\` configuration that allows you to define which parts of your function’s input should be treated as attachments, specifying their MIME types and data. This function is called whenever you invoke the traceable-wrapped function, and it enables you to extract binary data (e.g., images, audio files) from your inputs and log them alongside other trace data.

Reviewer: nit: This could be more concise
```suggestion
In the TypeScript SDK, the \`extractAttachments\` function is an optional parameter in the \`traceable\` configuration. When the traceable-wrapped function is invoked, it extracts binary data (e.g., images, audio files) from your inputs and logs them alongside other trace data, specifying their MIME types.
```
